Crypto/OTP.hs

```diff
@@ -26,6 +26,43 @@
 -- >>>
 -- >>> let getOTPTime = getPOSIXTime >>= \t -> return (floor t :: OTPTime)
 --
+-- == Example Usage: Integration with Google Authenticator
+--
+-- Google Authenticator imposes a few restrictions on TOTP, such as only supporting 6 digit codes and 30 second time windows. An implementation would look something like:
+--
+-- 1. Generate a secret key and store it in a database, marking the code as unverified.
+-- 2. Return the encoded secret key to the client.
+-- 3. The client should render a QR code with the data: @otpauth:\/\/totp\/URL_ENCODED_COMPANY_NAME?secret=ENCODED_SECRET&issuer=URL_ENCODED_COMPANY_NAME@
+-- 4. The user should scan the QR code using Google Authenticator, then submit their 6 digit code ('OTP').
+-- 5. Use 'totpVerify' with 'defaultTOTPParams' to validate the code, then mark the code as verified.
+-- 6. Future login attempts should require a new 'OTP' code.
+--
+-- This code demonstrates how to generate the secret key:
+--
+-- > import Control.Monad.IO.Class
+-- > import qualified Data.Text as T
+-- > import qualified Data.Text.Encoding as TE
+-- > import Data.ByteArray (ByteArrayAccess) -- memory package
+-- > import Data.ByteArray.Encoding (convertToBase, Base(Base32))
+-- > import Crypto.Random (getSystemDRG, randomBytesGenerate)
+-- >
+-- > newtype TOTPSecretKey = TOTPSecretKey ByteString
+-- >     deriving (Show, ByteArrayAccess)
+-- > 
+-- > newtype GoogleAuthenticatorEncodedKey = GoogleAuthenticatorEncodedKey Text
+-- >   deriving (Show)
+-- >
+-- > -- Google Authenticator accepts arbitrary length secrets, using 80 bits for its own services.
+-- > -- RFC 4226 recommends a 160 bit (20 byte) secret key https://tools.ietf.org/html/rfc4226#section-4
+-- > randomTOTPSecretKey :: MonadIO m => m (TOTPSecretKey, GoogleAuthenticatorEncodedKey)
+-- > randomTOTPSecretKey = do
+-- >   key <- liftIO $ fst . randomBytesGenerate 20 <$> getSystemDRG
+-- >   return (TOTPSecretKey key, googleAuthenticatorEncode key)
+-- >
+-- > googleAuthenticatorEncode :: ByteString -> GoogleAuthenticatorEncodedKey
+-- > googleAuthenticatorEncode bs = GoogleAuthenticatorEncodedKey $ stripPadding $ TE.decodeUtf8 $ convertToBase Base32 bs
+-- >   where
+-- >     stripPadding t = T.dropWhileEnd (== '=') t
 
```

Comment:
My main concern with the code is that it is quite long for what it is showing (essentially just generating and encoding some bytes, which isn't specific to OTP). You could perhaps summarise by saying something like:

> Generate a random 20 byte key using [`getRandomBytes 20`](http://hackage.haskell.org/package/cryptonite-0.25/docs/Crypto-Random.html#v:getRandomBytes), store the key in the database and return the base-32 encoded key to the user for entry into the Google Authenticator application. The key can be encoded using [`convertToBase Base32`](http://hackage.haskell.org/package/memory-0.14.16/docs/Data-ByteArray-Encoding.html#v:convertToBase) from the `memory` package.

I don't really mind about the size of docs in the code, but I think an extended code sample with imports should be compilable (which I don't think this is), and show how to use the functionality of the module in question. For example, sample code to generate a sequence of bytes should probably be in the `Crypto.Random` module and could then be linked from elsewhere.